### ISSUE TITLE
wrong implementation of "torch.optim.lr_scheduler.CosineAnnealingLR"

### ISSUE THREAD

**Issue by Xiejiu** [pytorch/pytorch#37916]
## 🐛 Bug

 I want to adjust the learning rate using a cosine annealing schedule at each float epoch rather than an int epoch.  I found that when calling the "torch.optim.lr_scheduler.CosineAnnealingLR", the output is wrong. 

Following codes are saved as '"try.py"
-------------------------
from torch.optim.lr_scheduler import _LRScheduler
import torch
import math

class CosineAnnealingLR(_LRScheduler):
    def __init__(self, optimizer, T_max, eta_min=0, last_epoch=-1):
        self.T_max = T_max
        self.eta_min = eta_min
        super(CosineAnnealingLR, self).__init__(optimizer, last_epoch)

    def get_lr(self):
        return [self.eta_min + (base_lr - self.eta_min) *
                (1 + math.cos(math.pi * self.last_epoch / self.T_max)) / 2
                for base_lr in self.base_lrs]

if __name__ == '__main__':
    v = torch.zeros(10)
    optim_1 = torch.optim.SGD([v], lr=0.05)
    optim_2 = torch.optim.SGD([v], lr=0.05)
    total_epoches = 5
    scheduler = CosineAnnealingLR(optim_1, T_max=total_epoches, eta_min=0, last_epoch=-1)
    cosine_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim_2, T_max=total_epoches, eta_min=0, last_epoch=-1)
            
    step = 1
    for epoch in range(0, 5):
        for tmp_step in range(0,5):
            current_epoch = epoch + tmp_step / 5
            scheduler.step(current_epoch)
            cosine_scheduler.step(current_epoch)
            print(current_epoch, optim_1.param_groups[0]['lr'])
            print(current_epoch, optim_2.param_groups[0]['lr'])
            step = step + 1  
------------------------    

## Expected output
optim_1 and optim_2 have the same results

## real output

0.0 0.05
0.0 0.05
0.2 0.04980286753286195
0.2 0.053086063235353066
0.4 0.04921457902821578
0.4 0.054153592884941676
0.6 0.04824441214720629
0.6 0.05308606323535307
0.8 0.04690766700109659
0.8 0.05000000000000001
1.0 0.04522542485937369
1.0 0.0452254248593737
1.2 0.04322421568553529
1.2 0.03925142497671765
1.4 0.040935599743717244
1.4 0.03264846827799254
1.6 0.038395669874474916
1.6 0.0259835233577752
1.8 0.03564448228912682
1.8 0.019744517204696613
2.0 0.032725424859373686
2.0 0.014287266867609881
2.2 0.029684532864643117
2.2 0.009811880589412713
2.4 0.02656976298823284
2.4 0.006368523812077529
2.6 0.023430237011767167
2.6 0.0038862721452675294
2.8 0.020315467135356886
2.8 0.0022149693017232894
3.0 0.017274575140626316
3.0 0.0011692026551594127
3.2 0.014355517710873185
3.2 0.0005654294612037747
3.4 0.01160433012552509
3.4 0.0002469510222357759
3.6 0.009064400256282757
3.6 9.553735662678324e-05
3.8 0.006775784314464717
3.8 3.186441730156198e-05
4.0 0.004774575140626317
4.0 8.807108335809691e-06
4.2 0.0030923329989034107
4.2 1.8971459114367982e-06
4.4 0.0017555878527937135
4.4 2.8701409569257634e-07
4.6 0.0007854209717842259
4.6 2.4869476587641312e-08
4.8 0.0001971324671380559
4.8 7.235444708132142e-10

## You can get the script and run it with:

 python try.py

## Environment

 - PyTorch Version (e.g., 1.0): 1.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.6
 - CUDA/cuDNN version: 9.0.176/7.2.1

cc @vincentqb

**Comment by vincentqb** [comment]
Please make sure you are using the latest version of pytorch (1.5.0). This has been fixed in #20527. I will close this issue, but please feel free to re-open if the problem persists with the latest version.
